File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/MapRouteProgressChangeListener.java
package com.mapbox.services.android.navigation.v5.route;

import android.location.Location;

import com.mapbox.services.android.navigation.v5.models.DirectionsRoute;
import com.mapbox.services.android.navigation.v5.navigation.NavigationMapRoute;
import com.mapbox.services.android.navigation.v5.routeprogress.ProgressChangeListener;
import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress;

import java.util.List;

/**
 * @deprecated this class is outdated, use {@link com.mapbox.services.android.navigation.ui.v5.route.MapRouteProgressChangeListener} instead.
 */
@Deprecated
public class MapRouteProgressChangeListener implements ProgressChangeListener {

  private final NavigationMapRoute mapRoute;

  public MapRouteProgressChangeListener(NavigationMapRoute mapRoute) {
    this.mapRoute = mapRoute;
  }

  @Override
  public void onProgressChange(Location location, RouteProgress routeProgress) {
    DirectionsRoute currentRoute = routeProgress.directionsRoute();
    List<DirectionsRoute> directionsRoutes = mapRoute.retrieveDirectionsRoutes();
    int primaryRouteIndex = mapRoute.retrievePrimaryRouteIndex();
    addNewRoute(currentRoute, directionsRoutes, primaryRouteIndex);
    mapRoute.addUpcomingManeuverArrow(routeProgress);
  }

  private void addNewRoute(DirectionsRoute currentRoute, List<DirectionsRoute> directionsRoutes,
                           int primaryRouteIndex) {
    if (isANewRoute(currentRoute, directionsRoutes, primaryRouteIndex)) {
      mapRoute.addRoute(currentRoute);
    }
  }

  private boolean isANewRoute(DirectionsRoute currentRoute, List<DirectionsRoute> directionsRoutes,
                              int primaryRouteIndex) {
    boolean noRoutes = directionsRoutes.isEmpty();
    return noRoutes || !currentRoute.equals(directionsRoutes.get(primaryRouteIndex));
  }
}

File: libandroid-navigation/src/main/java/org/maplibre/navigation/android/navigation/v5/route/MapRouteProgressChangeListener.kt
package org.maplibre.navigation.android.navigation.v5.route

import android.location.Location
import org.maplibre.navigation.android.navigation.v5.models.DirectionsRoute
import org.maplibre.navigation.android.navigation.v5.navigation.NavigationMapRoute
import org.maplibre.navigation.android.navigation.v5.routeprogress.ProgressChangeListener
import org.maplibre.navigation.android.navigation.v5.routeprogress.RouteProgress

@Deprecated("Use in ui package instead")
class MapRouteProgressChangeListener(private val mapRoute: NavigationMapRoute) : ProgressChangeListener {
    
    override fun onProgressChange(location: Location, routeProgress: RouteProgress) {
        val currentRoute = routeProgress.directionsRoute
        val directionsRoutes = mapRoute.retrieveDirectionsRoutes()
        val primaryRouteIndex = mapRoute.retrievePrimaryRouteIndex()
        addNewRoute(currentRoute, directionsRoutes, primaryRouteIndex)
        mapRoute.addUpcomingManeuverArrow(routeProgress)
    }

    private fun addNewRoute(
        currentRoute: DirectionsRoute,
        directionsRoutes: List<DirectionsRoute>,
        primaryRouteIndex: Int
    ) {
        if (isANewRoute(currentRoute, directionsRoutes, primaryRouteIndex)) {
            mapRoute.addRoute(currentRoute)
        }
    }

    private fun isANewRoute(
        currentRoute: DirectionsRoute,
        directionsRoutes: List<DirectionsRoute>,
        primaryRouteIndex: Int
    ): Boolean {
        val noRoutes = directionsRoutes.isEmpty()
        return noRoutes || currentRoute != directionsRoutes[primaryRouteIndex]
    }
}
